In [1]:
# 导入sklearn内部的手写体数据集，完整数据在http://archive.ics.uci.edu/ml/index.php查阅
from sklearn.datasets import load_digits

digits = load_digits()
digits.data.shape

(1797, 64)

传统机器学习，只能识别将2D的图片像素矩阵首尾相连变成1D的像素特征向量，这样会损失一些数据本身的结构信息，但用深度学习技术可以处理结构化数据信息

数据有1797条，每条由8*8=64像素矩阵表示

## 切分数据

In [2]:
from sklearn.model_selection import train_test_split

In [3]:
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                        test_size = 0.25, random_state = 33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(1347, 64) (450, 64) (1347,) (450,)


## 使用SVM进行识别

In [4]:
from sklearn.preprocessing import StandardScaler
# 将数据标准化
ss = StandardScaler()
X_train = ss.fit_transform(X_train)
X_test = ss.transform(X_test)

In [5]:
# 导入基于线性假设的支持向量机
from sklearn.svm import LinearSVC
lsvc = LinearSVC()
# 训练模型
lsvc.fit(X_train, y_train)
# 存储预测结果
y_predict = lsvc.predict(X_test)

## 对模型效果评估

In [6]:
print('The accuracy of linearSVC:', lsvc.score(X_test, y_test))

The accuracy of linearSVC: 0.9533333333333334


In [7]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict, target_names = digits.target_names.astype(str)))

             precision    recall  f1-score   support

          0       0.92      1.00      0.96        35
          1       0.96      0.98      0.97        54
          2       0.98      1.00      0.99        44
          3       0.93      0.93      0.93        46
          4       0.97      1.00      0.99        35
          5       0.94      0.94      0.94        48
          6       0.96      0.98      0.97        51
          7       0.92      1.00      0.96        35
          8       0.98      0.84      0.91        58
          9       0.95      0.91      0.93        44

avg / total       0.95      0.95      0.95       450



召回率，准确率，f1适用于二分类任务，在多分类中，只需将其他类别看为负样本即可，如评估0时，将1-9看为负样本

支持向量机的特点是在小规模数据集能取得较好效果，不适用于大规模数据